homeassistant/components/squeezebox/media_player.py

```diff
@@ -126,18 +141,75 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
     lms = Server(async_get_clientsession(hass), host, port, username, password)
     known_servers.add(ipaddr)
 
-    players = await lms.async_get_players()
-    if players is None:
-        raise PlatformNotReady
-    media_players = []
-    for player in players:
-        media_players.append(SqueezeBoxDevice(player))
+    async def _discovery(now=None):
+        """Discover squeezebox players by polling server."""
+
+        async def _discovered_player(player):
+            """Handle a (re)discovered player."""
+            entity = next(
+                (
+                    known
+                    for known in known_players
+                    if known.unique_id == player.player_id
+                ),
+                None,
+            )
+            if entity and not entity.available:
+                # check if previously unavailable player has connected
+                await player.async_update()
+                entity.available = player.connected
```

Comment:
Can we avoid having to set `available` from outside the object and have `async_update` check `self._player.connected` instead?

Reply:
Unfortunately, no. This is a problem with how Logitech Media Server handles disconnected players. For a short period of time (a few minutes), `connected` is false. After that period of time elapses, the server forgets completely about the player, and abruptly disconnects if you attempt to query information about it.

My workaround is stop trying to query disconnected players by having `async_update` skip them. If they are re-discovered, I mark them `connected` so we start updating them again. I check `player.connected` just in case we have rediscovered a player during the limited time it is found by the server but in the disconnected state.

Reply:
We want to avoid accessing the entity from outside. We can use our dispatch helper to signal the an entity method from outside.

Reply:
Ok. I think this got merged before your comment, but I can create a new PR to address this and your other comments below.